### DISCUSSION TITLE
Unable to use regex expression

### DISCUSSION THREAD

**Discussion by TheFern2** [micropython/micropython#10299] [Unix & Windows] [answered]
I have search the webz and I think maybe the problem is string escaping but I am not 100% sure since is working in python3.10 just fine. I have read some posts about another RE library for micropython but I was unable to figure out how to install it with mip (re-pcre) maybe it doesn't exists anymore. https://forum.micropython.org/viewtopic.php?t=5398

I have strings like this where it should match square brackets:
```
BaseBOOLArray[2]
MyTag[0].UDTMember[4]
```

```python
array_pattern = re.compile('\[([\d]|[,]|[\s])*\]')
index = array_pattern.search(tag).group(0)
```

Can anyone point me into what I am doing wrong. Running micropython v1.19 latest on main branch unix port.

**Comment by scruss** [comment, marked as answer]
MicroPython's [u]re module is very small. The re-pcre module is old and is no longer in micropython-lib. MicroPython is a subset of an older Python than 3.10.

What are you looking for? `\[.+\]` will match square bracket groups (or `\[[^\]]+\]` if you're cautious). I don't think we have PCRE's pipe matching at all.

**Reply by TheFern2** [reply]
Nvm for whatever reason \d wasn't working in mpy. I changed to 0-9 and it worked. Thanks for the response!

**Reply by TheFern2** [reply]
Btw what is a good way to search for libraries available for mpy, pypy site?

**Reply by git-n-pissed** [reply]
> Btw what is a good way to search for libraries available for mpy, pypy site?

The [lib](https://github.com/micropython/micropython-lib/tree/038b4ac6572d0d5b4c18148dc7d7fdc026369fa4) directory of the MicroPython project has all the libraries that have been officially ported over.  You can add them to your project by including/requiring them in your manifest.  See https://docs.micropython.org/en/latest/reference/manifest.html

For non-official libraries, I don't know if there is a better way to find them than doing a search on Google, GitHub (starting in these discussions), or the old MicroPython forum.  Maybe a good idea for someone to start keeping an organized list though, there are quite a few!

**Reply by TheFern2** [reply]
Ah ok thanks @git-n-pissed I am brand new to the micropython world, and just trying to see what's out there I am a collaborator on a python lib and I am working on the port for micropython at the moment.

**Reply by mattytrentini** [reply]
A couple us also maintain the [Standard Library Coverage](https://github.com/micropython/micropython/wiki/Standard-Library-Coverage) wiki page. It attempts to indicate the status of _all_ the Python libraries implemented for MicroPython.

(Happy to have more contributors too!)

**Reply by mattytrentini** [reply]
I'm also confused as to why your regex didn't work - it could be simplified but I believe all the match tokens you've used are valid for MicroPython. `\d` and `|` should work, but maybe the structure of that particular regex is causing grief..?

I'll try and dig into it a little further - mostly because I'd like to understand and document if there's a difference here - but I'm glad you have a workaround.

**Reply by TheFern2** [reply]
Yes, our library pylogix had that specific regex for ages, now that I'm making sure it works in micropython I found that \d wasn't working, someone in our discord suggested to use 0-9 and to my surprise it worked. Strange indeed. Lemme know if you find different results.

**Comment by peterhinch** [comment]
For users seeking unofficial libraries see [awesome micropython](https://awesome-micropython.com/).

**Comment by jimmo** [comment]
@TheFern2 Sorry about the delay responding...

The issue here is that Python supports using named character classes (e.g. `\d`) inside character classes (e.g. `[0-9]`). I guess the idea is that you can write `[x\d]` to support "x or digit" as a shorthand for `[x0-9]`.

You've used `[\d]` but for MicroPython you need to just write `\d` (and it would behave equivalently in Python).

**Reply by jimmo** [reply]
I just realised this is a known issue (with a PR to implement it) -- see https://github.com/micropython/micropython/pull/7922 and https://github.com/micropython/micropython/issues/7920

**Reply by TheFern2** [reply]
Thanks for the response. As it turned out one of our discord community members was a regex expert and he contributed to the regex we ended up using which is compatible with python and mpy.
